### PR TITLE
Efficient burn in for sampling

### PR DESCRIPTION
**Pull request by vgnecula**
Modified the sampling.hpp to solve:  Inefficient handling of burn in phase in sampling functions #180.

Handling burn in phase without storing the points required for this phase, for each of the sampling methods presented.

Cleared my branch non-related developments in order to emphasize just the sampling.hpp code.

### CODE REVIEW

File: include/sampling/sampling.hpp [hunk 1]
```diff
@@ -21,45 +21,37 @@
...
+            walk.apply(P, p, walk_len, rng);  
+        }
     }
-    RandomPointGenerator::apply(P, p, rnum, walk_len, randPoints,
-                                push_back_policy, rng);
-
 
+    for (unsigned int i = 0; i < rnum; ++i) {
+        walk.apply(P, p, walk_len, rng); 
+        randPoints.push_back(p);
```
Comment (TolisChal) [inline include/sampling/sampling.hpp]: Could you please reserve memory equal to rnum points before the push_backs? (i.e., randPoints.reserve(rnum);)

File: include/sampling/sampling.hpp [hunk 2]
```diff
@@ -68,169 +60,157 @@ template <
...
+        for (unsigned int i = 0; i < nburns; ++i) {
+            walk.apply(P, p, walk_len, rng);  
+        }
+    }
+
+    randPoints.clear();
+
+    for (unsigned int i = 0; i < rnum; ++i) {
+        walk.apply(P, p, walk_len, rng);  
+        randPoints.push_back(p);
```
Comment (TolisChal) [inline include/sampling/sampling.hpp]: same as before

File: include/sampling/sampling.hpp [hunk 3]
```diff
@@ -68,169 +60,157 @@ template <
...
+        }
     }
-    RandomPointGenerator::apply(P, p, a, rnum, walk_len, randPoints,
-                                push_back_policy, rng);
 
+    randPoints.clear(); 
 
+    for (unsigned int i = 0; i < rnum; ++i) {
+        walk.apply(P, p, a, walk_len, rng); 
+        randPoints.push_back(p);
```
Comment (TolisChal) [inline include/sampling/sampling.hpp]: same as before

File: include/sampling/sampling.hpp [hunk 4]
```diff
@@ -68,169 +60,157 @@ template <
...
+        }
+    }
+
+    
+    randPoints.clear();
+
+    
+    for (unsigned int i = 0; i < rnum; ++i) {
+        walk.apply(P, p, a, walk_len, rng); 
+        randPoints.push_back(p);
```
Comment (TolisChal) [inline include/sampling/sampling.hpp]: same as before

File: include/sampling/sampling.hpp [hunk 5]
```diff
@@ -239,65 +219,40 @@ void logconcave_sampling(PointList &randPoints,
...
-    typedef LogconcaveRandomPointGenerator<walk> RandomPointGenerator;
-    
     if (nburns > 0) {
-        RandomPointGenerator::apply(nburns, walk_len, randPoints,
-                                push_back_policy, rng, logconcave_walk);
+        for (unsigned int i = 0; i < nburns; ++i) {
+            walk.apply(P, p, walk_len, rng); 
+        }
     }
-    logconcave_walk.disable_adaptive();
```
Comment (TolisChal) [inline include/sampling/sampling.hpp:280]: Why did you delete this line?

File: include/sampling/sampling.hpp [hunk 6]
```diff
@@ -309,54 +264,38 @@ void crhmc_sampling(PointList &randPoints,
...
-                  NegativeLogprobFunctor,
-                  Solver
-          > walk;
-  typedef typename WalkTypePolicy::template parameters
-          <
-                  NT,
-                  NegativeGradientFunctor
-          > walk_params;
-  Point p = Point(problem.center);
-  problem.options.simdLen=simdLen;
```
Comment (TolisChal) [inline include/sampling/sampling.hpp:340]: This line it is deleted and not re-written. Please be sure that your changes modify only how volesti stores the points and don't affect anything else.

**Comment by TolisChal** [comment]
@vgnecula thanks for this PR!

A general comment is, when you want to change a particular part of the code, just change this one without affecting anything else. Then, it's easier to review the code and be sure about the changes.

File: include/sampling/sampling.hpp [hunk 7]
```diff
@@ -68,169 +60,157 @@ template <
...
     if (nburns > 0) {
-        RandomPointGenerator::apply(P, p, nburns, walk_len, randPoints,
-                                    push_back_policy, rng, WalkType.param);
-        randPoints.clear();
+        for (unsigned int i = 0; i < nburns; ++i) {
+            walk.apply(P, p, walk_len, rng);  
+        }
+    }
+
+    randPoints.clear();
```
Comment (vfisikop) [inline include/sampling/sampling.hpp]: clear is not needed here.

File: include/sampling/sampling.hpp [hunk 8]
```diff
@@ -68,169 +60,157 @@ template <
...
 
-template
-<
-        typename WalkTypePolicy,
-        typename PointList,
-        typename Polytope,
-        typename RandomNumberGenerator,
-        typename Point
+
+template <
```
Comment (vfisikop) [inline include/sampling/sampling.hpp]: I would prefer to change style at a separate PR.

File: include/sampling/sampling.hpp [hunk 9]
```diff
@@ -68,169 +60,157 @@ template <
...
+        for (unsigned int i = 0; i < nburns; ++i){
+            walk.apply(P, p1, p2, walk_len, rng);
+        }
     }
-    unsigned int n = rnum / 2;
-    BoundaryRandomPointGenerator::apply(P, p, rnum / 2, walk_len,
-                                        randPoints, push_back_policy, rng);
 
+    
+    randPoints.clear();
```
Comment (vfisikop) [inline include/sampling/sampling.hpp]: is this needed?

File: include/sampling/sampling.hpp [hunk 10]
```diff
@@ -68,169 +60,157 @@ template <
...
+
+    for (unsigned int i = 0; i < rnum; ++i) {
```
Comment (vfisikop) [inline include/sampling/sampling.hpp]: there is a mistake here, it was rnum/2 before

File: include/sampling/sampling.hpp [hunk 11]
```diff
@@ -68,169 +60,157 @@ template <
...
+        walk.apply(P, p1, p2, walk_len, rng); 
+        randPoints.push_back(p1);
```
Comment (vfisikop) [inline include/sampling/sampling.hpp]: should we also push p2?

File: include/sampling/sampling.hpp (hunk 9 above)
Comment (vfisikop) [reply include/sampling/sampling.hpp]: similar in all walks

File: include/sampling/sampling.hpp [hunk 12]
```diff
@@ -68,169 +60,157 @@ template <
...
-                    RandomNumberGenerator
-            > walk;
-
-    //RandomNumberGenerator rng(P.dimension());
-    PushBackWalkPolicy push_back_policy;
+    
+    typename WalkTypePolicy::template Walk<Polytope, RandomNumberGenerator> walk(P, starting_point, a, rng);
 
-    Point p = starting_point;
+    Point p = starting_point;
```
Comment (vfisikop) [inline include/sampling/sampling.hpp]: please do not add trailing whitespaces (there are more cases like this in this PR)

File: include/sampling/sampling.hpp [hunk 13]
```diff
@@ -239,65 +219,40 @@ void logconcave_sampling(PointList &randPoints,
...
-                    Solver
-            > walk;
-
-    typedef typename WalkTypePolicy::template parameters
-            <
-                    NT,
-                    NegativeGradientFunctor
-            > walk_params;
-
-    // Initialize random walk parameters
```
Comment (vfisikop) [inline include/sampling/sampling.hpp:260]: there is probably some mistake deleting this

File: include/sampling/sampling.hpp [hunk 14]
```diff
@@ -309,54 +264,38 @@ void crhmc_sampling(PointList &randPoints,
...
+    for(unsigned int i = 0; i < nburns; ++i) {
+        crhmc_walk.apply(rng, walk_len);
+    }
+
+    for(unsigned int i = 0; i < rnum; ++i) {
+        crhmc_walk.apply(rng, walk_len);
+        randPoints.push_back(crhmc_walk.getPoint()); 
+    }
 }
+
```
Comment (vfisikop) [inline include/sampling/sampling.hpp:345]: too many extra lines

File: include/sampling/sampling.hpp [hunk 15]
```diff
@@ -434,169 +373,162 @@ crhmc_sampling <
...
+    PointList &randPoints,
+    Polytope &P,
+    RandomNumberGenerator &rng,
+    const unsigned int &walk_len,
+    const unsigned int &rnum,
+    const Point &c,
+    const NT &a,
+    Point &starting_point, 
+    unsigned int const& nburns
+) {
```
Comment (vfisikop) [inline include/sampling/sampling.hpp]: looks like a new style of code, I suggest using the style of most of the functions in volesti

**Review by vfisikop — COMMENTED** [review]
Thanks. I suggest as @TolisChal also suggested to focus on the issue of the burn in phase and leave style changes for future PR.

**Comment by vfisikop** [comment]
Also it seems that you are deleting (by mistake) some crucial parts of the code in sampling.hpp.

File: include/sampling/sampling.hpp (hunk 1 above)
Comment (vgnecula) [reply include/sampling/sampling.hpp]: I am not able to do that directly using randPoints.reserve(rnum); as randPoints is an instance of an std::list and not std::vector. Thus it does not support this type of command. Am I missing something?

Comment (vgnecula) [reply include/sampling/sampling.hpp]: I thought about finding a workaround for reserving space from the first time, however as I figured out randPoints was a list, I decided that it is not the case anymore, as I can't get continuous storage anyway.

File: include/sampling/sampling.hpp (hunk 9 above)
Comment (vgnecula) [reply include/sampling/sampling.hpp]: Initially, I did not have this function. However, I thought that it might be a good practice to ensure that everything is cleared before performing the actual sampling. Besides, this operation should be done in constant time (correct me if I am wrong), especially in the case in which we expect randPoints to be empty. Consequently, I thought that there would be no problem in adding it just for safety.

File: include/sampling/sampling.hpp (hunk 1 above)
Comment (vfisikop) [reply include/sampling/sampling.hpp]: `randPoints` is templated (i.e. it can be either a std::vector or a std::list or any container that supports `push_back`), thus you can create a specialized version of this function that reserves memory only for the case of a vector but I propose not to do anything in this PR. We should think if we need to want to keep it as it is or restrict `randPoints` to std::list or std::vector.
@TolisChal do we need random access in samples? If not should we restrict to std::list here?

**Comment by vfisikop** [comment]
Side comment: there is no need to open a new PR every time you want to change something in your branch, just update your branch!

File: include/sampling/sampling.hpp (hunk 11 above)
Comment (vgnecula) [reply include/sampling/sampling.hpp:122]: From what I understood regarding how boundary sampling works, the p2 is a kind of dummy point used for processing the walk and updating p1. Thus, I think that we only need to push p1.

File: include/sampling/sampling.hpp (hunk 13 above)
Comment (vgnecula) [reply include/sampling/sampling.hpp:260]: I do not use walk_params anymore.

**Comment by vgnecula** [comment]
@TolisChal @vfisikop just commited a new version, which adresses all your comments (find it as - minor probem with CRHMC solved - the last version of my new modifications). On my end it passes all tests. However, it takes way too much time to process crhmc_test_polytope_sampling. Besides, as you will probably see, I modified some functions, as I did not like some details that I implemented before.

File: include/sampling/sampling.hpp [hunk 16]
```diff
@@ -22,215 +22,193 @@
...
-                   RandomNumberGenerator &rng,
-                   const unsigned int &walk_len,
-                   const unsigned int &rnum,
-                   const Point &starting_point,
-                   unsigned int const& nburns)
+                      Polytope &P,
+                      RandomNumberGenerator &rng,
+                      const unsigned int &walk_len,
+                      const unsigned int &rnum,
+                      Point &starting_point,
```
Comment (vfisikop) [inline include/sampling/sampling.hpp]: please avoid using trailing whitespaces.

Comment (vfisikop) [reply include/sampling/sampling.hpp]: there are similar cases below

File: include/sampling/sampling.hpp [hunk 17]
```diff
@@ -22,215 +22,193 @@
...
+    Point current_point = starting_point;
+    Point dummy_point(P.dimension());
 
-    typedef BoundaryRandomPointGenerator <walk> BoundaryRandomPointGenerator;
-    if (nburns > 0) {
-        BoundaryRandomPointGenerator::apply(P, p, nburns, walk_len,
-                                            randPoints, push_back_policy, rng);
-        randPoints.clear();
+    for (unsigned int i = 0; i < nburns; ++i) {
+        walk.apply(P, current_point, dummy_point, walk_len, rng);
```
Comment (vfisikop) [inline include/sampling/sampling.hpp:125]: I think there is not reason to keep only one point, they are both on the boundary, so I propose to keep both. @TolisChal what do you think?

File: include/sampling/sampling.hpp [hunk 18]
```diff
@@ -239,65 +217,48 @@ void logconcave_sampling(PointList &randPoints,
...
-
-    typedef typename WalkTypePolicy::template parameters
-            <
-                    NT,
-                    NegativeGradientFunctor
-            > walk_params;
-
-    // Initialize random walk parameters
-    unsigned int dim = starting_point.dimension();
-    walk_params params(F, dim);
```
Comment (vfisikop) [inline include/sampling/sampling.hpp:262]: how you pass the parameters to the walk?

**Review by vfisikop — COMMENTED** [review]
I reviewed part of it, but it is difficult to continue. Could you please separate your commits such that there is one commit with the changes to burn in feasibility and another one with the style changes?

File: include/sampling/sampling.hpp (hunk 16 above)
Comment (vgnecula) [reply include/sampling/sampling.hpp]: Thank you for taking the time to review my new Commits!

I just wanted to be sure I understand what you mean by style changes.

So, I should have a first commit that for the lines where I did not change anything regarding the algorithm logic - the lines remain the same, while the second commit adds my proposed style changes.

I saw that there are some small differences in the ways some things are written and I tried to bring them all to the same writing style (that is why I made those modifications). Sorry for the white spaces, I missed some of them.

Comment (vfisikop) [reply include/sampling/sampling.hpp]: Yes, exactly, we want one commit to fix the burn in phase that does not change any coding style (i.e. indentation etc) and another commit that contains all coding styles. Also please rebase (see `git rebase` documentation) your branch to only have those two commits and not the current ones (currently 4).

File: include/sampling/sampling.hpp (hunk 18 above)
Comment (vgnecula) [reply include/sampling/sampling.hpp:262]: @vfisikop Thank you for pointing this out! So previously I was only relying on the assumption that the walk object will auto set the params. However after looking into how the langevin_walk is implemented, I saw that this is not the case. Thus, now I am setting the params using the set_params method. Let me know if this makes sense.

File: include/sampling/sampling.hpp (hunk 16 above)
Comment (vgnecula) [reply include/sampling/sampling.hpp]: @vfisikop should I push the "style-changes" commit? Or should I wait until the current one gets reviewed? Thank you!

Comment (vfisikop) [reply include/sampling/sampling.hpp]: The PR should consist of two commits, one with the fixes and one with the style changes. So those could be reviewed independently.

Comment (vfisikop) [reply include/sampling/sampling.hpp]: Having a look at this code again I see there are some problems with this interface, e.g. there is an issue of what walk should the user use for a particular distribution. You can have a look at this draft PR https://github.com/GeomScale/volesti/pull/300 for an idea of changing this to a new interface.

Comment (vgnecula) [reply include/sampling/sampling.hpp]: I just started to look into it more. Could you please elaborate a bit more on what you think I should do? I want to be sure I fully understood what you meant.

File: include/sampling/sampling.hpp (hunk 17 above)
Comment (TolisChal) [reply include/sampling/sampling.hpp:125]: Yes, we could do that, however keeping only one would be closer to a uniform sample (just intuitively).
A general comment from my side is that we need to delete this method in the future and implement Shake and Bake.
